docs/proposals/20200602-machine-deletion-phase-hooks.md

```diff
@@ -0,0 +1,289 @@
+
+##### hook-name
+Each hook should have a unique and descriptive name that describes in 1-3 words what the intent/reason for the hook is.  Each hook name should be unique and managed by a single entity.
+
+##### owner (Optional)
+Some information about who created or is otherwise in charge of managing the annotation.  This might be a controller or a username to indicate an administrator applied the hook directly.
+
+##### Annotation Examples
+pre-drain.hook.machine.cluster-api.x-k8s.io/migrate-important-app: my-app-migration-controller
+
+pre-term.hook.machine.cluster-api.x-k8s.io/preserve-instance: my-instance-replacement-controller
+
+pre-term.hook.machine.cluster-api.x-k8s.io/never-delete: cluster-admin
+
+#### Changes to machine-controller
+The machine-controller should check for the existence of 1 or more hooks at
+specific points (lifecycle-points) during reconciliation.  If a hook matching
+the lifecycle-point is discovered, the machine-controller should stop
+reconciling the machine.
+
+An example of where the pre-drain lifecycle-point might be implemented:
+https://github.com/kubernetes-sigs/cluster-api/blob/30c377c0964efc789ab2f3f7361eb323003a7759/controllers/machine_controller.go#L270
+
+##### Reconciliation
+When an Owning Controller updates the machine, reconciliation will be
+triggered, and the machine will continue reconciling as normal, unless another
+hook is still present; there is no need to 'fail' the reconciliation to
+enforce requeuing.
+
+When all hooks for a given lifecycle-point are removed, reconciliation
+will continue as normal.
+
+##### Hook failure
+The machine-controller should not timeout or otherwise consider the lifecycle
+hook as 'failed.'  Only the Owning Controller may decide to remove a
+particular lifecycle hook to allow the machine-controller to progress past the
+corresponding lifecycle-point.
+
+##### Hook ordering
+The machine-controller will not attempt to enforce any ordering of hooks.
+
+Owning Controllers may devise a dependency-based enforcement mechanism
+via whatever means Owning Controllers determine.  Examples could be
+using CRDs external to the machine-api, gRPC communications, or
+additional annotations on the machine or other objects.
+
+#### Owning Controller Design
+Owning Controller is the component that manages a particular lifecycle hook.
+
+##### Owning Controllers must
+* Watch machine objects and determine when an appropriate action must be taken.
+* After completing the desired hook action, remove the hook annotation.
+##### Owning Controllers may
+* Watch machine objects and add a hook annotation as desired by the operator
+* After completing a hook, set an addition annotation that indicates this
+controller has finished for dependency ordering, or perform some other action
+to signal other Owning Controllers that they may proceed.
+
+#### Determining when to take action
+
+An Owning Controller should watch machines and determine when is the
+best time to take action.
```

Comment:
Should we implement some utils so we can ensure a common interpretation of when a lifecycle-point should be triggered?

Reply:
Right now, I think all we have that is native to the machine object is DeletionTimestamp.  Any other more specific case would need to be deduced by the implementer.  If we add more phases/sub-phases/annotations that better indicate a more specific machine state, we could then capture the need to provide such a util (either in this proposal, or some new future one).

Reply:
> Any other more specific case would need to be deduced by the implementer

this is kind of fragile, but I'm fine also in providing detailed documentation if we don't want to invest in a util

Reply:
One problem I see with a util is that it will be versioned at a given point in time only. If additional hook points are added later and not all external implementations are using the same version of the util you may end up with different semantics across them.

I think it might be better to leverage Conditions for this, so a hook doesn't trigger unless both:
- DeletionTimestamp is set
- Condition indicating lifecycle point is present and false (Machine controller can set it to true when there are no longer any hooks defined)

That would allow for triggering hooks only after the intended point in time as well as enforcing ordering between hooks at different lifecycle points.

Reply:
Condition for each lifecycle point makes sense to me.  It's still going to require knowledge on both ends of the equation.  If a newer owning controller is deployed onto an older cluster-api deployment, that will likely not work as designed.  Having all the points in a dictionary/list in a true/false format would at least allow owning controllers to validate that the point they intend to consume is functional on a machine.

> Condition indicating lifecycle point is present and false (Machine controller can set it to true when there are no longer any hooks defined)

I would say, setting to true when we reach that point.  So, I want to wait for the pre-drain.delete point, and when pre-drain.delete is set to true, I do my thing.  My thinking is more running hooks *at* at lifecycle point rather than before/after.

Anyway, this discussion really merits it's own CAEP design, I think.

Reply:
DeletionTimestamp set + Cluster/Machine not paused = ok to pre-drain

When is it ok to pre-terminate? How about we add a single new condition to Machine - NodeDrained - and set it to true after draining is done? Then we'd have:

DeletionTimestamp set + Cluster/Machine not paused + NodeDrained=True = ok to pre-terminate

Reply:
> When is it ok to pre-terminate?

So, I think you're looking at it from the reverse of the intended perspective.  pre-terminate *blocks* the machine controller from terminating the instance, it's not necessarily prescriptive of *when* the hook should be fulfilled, only that the hook blocks termination.

For instance, I might want to start an action immediately after the machine is marked deleted, and I might not want to block drain from happening, but I *do* want to prevent the instance from being terminated.

I agree, having some indication of the machine being drained might be desirable, but that is like a whole CAEP in itself and I don't think we need to incorporate those details here.

Reply:
You're right, I was looking at it the wrong way. I filed #3365 for the drained condition.

I am file to resolve this thread - WDYT @fabriziopandini @detiber?